### ISSUE TITLE
Amazon ECR Lifecycle Policies at cloud platform level

### ISSUE THREAD

**Issue by zheileman** [ministryofjustice/cloud-platform#3537]
Hello,

As services grow in the cloud platform and teams push many images to their ECRs with each deploy, and to different environments (usually the pattern for most teams is to have 1 ECR per service, where images for dev, staging, production, etc. are pushed) the number of images that are not relevant grow, costing extra money I suppose.

Many teams have been crafting ad-hoc solutions with bash scripts, AWS CLI, cron jobs, GitHub actions, etc. Other teams doesn't have any lifecycle management just yet, with pushed images growing without control (the default quota is 10,000 images per repository if it has not been changed).

Would this be something that is better to be solved by cloud platforms, so teams don't need to worry about, knowing their old ECR images are purged automatically by the platform? A global solution that applies to all ECR repos in the cloud platform our of the box?

Of course there are some gotchas:

* No matter how old, an image in use shouldn't be deleted. Many services not in active development, or in maintenance mode, will push/deploy very sparsely. Or even not push for many months.

* Also, even if the current deployed image is not deleted, it would be good to also maintain the previous one or two (or more, configurable?) to be able to do rollbacks even if that previous image(s) was pushed many months ago.

Amazon provides some lifecycle policies, according to this link, but I don't think this allows the level of management required to solve the above gotchas:

https://aws.amazon.com/blogs/compute/clean-up-your-container-images-with-amazon-ecr-lifecycle-policies/

Is this something cloud platform might be able to solve for service teams? If not, are there any policies available that teams can configure in their ECRs? 

Thank you!

**Comment by zheileman** [comment]
Hello,

As a possible approach to this, I've created a GitHub Action that can be used in a GitHub workflow.

I've been using it in one of my own services and ECR repo, and although it works, I might be missing something here, so it may be worth if someone from CP team can take a look.

The action is here, and I've added some documentation of how it works and how to use it:
https://github.com/ministryofjustice/ecr-images-cleaner-action

It is based on code originally used by other teams to do their cleanups, with some modifications and tweaks. I'm still in the process of improving it and then propagate it to other services I own.

It works with the secrets that the pipeline inject in the GitHub repositories. So is easy to use for teams using already GitHub workflows for their deploys.

Thanks!
